### PR TITLE
8301684: Fix test code to not get finalizer deprecation warnings

### PR DESCRIPTION
**Pull request by afshin-zafari**
To replace the finalizer use in the code, the `Cleaner` approach is used as stated in [Oracle doc on deprecated finalize() method](https://docs.oracle.com/javase/10/docs/api/java/lang/Object.html#finalize()).
Briefly:
1.  An instance of `Cleaner` (`java.lang.ref`) is created.
2. Using the `Cleaner`, an object is registered with a `Runnable` callback that is notified when the object is no longer reachable (GC'ed).
3. Write code in the callback to do other cleanings.

```java
Cleaner c = new Cleaner();
Cleanable cleanable = c.register(an_obj, a_runnable);
...
  //JVM notifies by calling the 
    a_runnable.run() {...}
```
### Tests
local: 
     vmTestbase/nsk/monitoring/stress/classload
      vmTestbase/nsk/jvmti/AttachOnDemand
      vmTestbase/nsk/jdi/ObjectReference/referringObjects
      vmTestbase/nsk/sysdict
mach5: tier 1-5

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301684](https://bugs.openjdk.org/browse/JDK-8301684): Fix test code to not get finalizer deprecation warnings

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12968/head:pull/12968` \
`$ git checkout pull/12968`

Update a local copy of the PR: \
`$ git checkout pull/12968` \
`$ git pull https://git.openjdk.org/jdk pull/12968/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12968`

View PR using the GUI difftool: \
`$ git pr show -t 12968`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12968.diff">https://git.openjdk.org/jdk/pull/12968.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java [hunk 1]
```diff
@@ -259,6 +314,12 @@ public boolean unloadClass(ExecutionController stresser) {
             }
         }
 
+        // Since customClassLoader object is set to null, we should also null the cleaning
+        // objbects to let them be created when new customClassLoader is created.
```
Comment (coleenp) [inline test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java:318]: typo: objbects

File: test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java [hunk 2]
```diff
@@ -92,6 +93,57 @@ public class ClassUnloader {
      */
     private CustomClassLoader customClassLoader = null;
 
+    /*
+     * This class is used as an alternative to deprecated finalize() method.
+     * It is used as a 'cleaner' whose run() method will be called when JVM
+     * makes an 'registered' object unreachable.
+     * Here it will be used for tracking the local instance of 'CustomClassLoader'.
+     */
+    private static class CustomClassLoaderCleaner implements Runnable{
```
Comment (coleenp) [inline test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java:102]: nit: can you add a space before open { in these lines?
Also use // comments ?

**Review by coleenp — COMMENTED** [review]
I don't know how cleaners work so here are some superficial comments.

File: test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java [hunk 3]
```diff
@@ -92,6 +93,57 @@ public class ClassUnloader {
...
+
+        /**
+         * Keep which ClassUnloader has to be notified.
+         */
+        public CustomClassLoaderCleaner(ClassUnloader class_unloader){
+            this.class_unloader = class_unloader;
+        }
+        @Override
+        public void run(){
+            if(class_unloader != null){
```
Comment (turbanoff) [inline test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java:114]:
```suggestion
            if (class_unloader != null){
```

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
This does not look quite right me - see comments below. You need to know that the Custom loader has become unreachable and only then check the unloaded classes.

There are numerous references to finalization still in the code that should be updated to a more suitable form.

I think things can be simplified/streamlined somewhat.

Thanks

File: test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java [hunk 4]
```diff
@@ -43,11 +44,14 @@
...
  * with especial <code>ClassUnloader.loadClass()</code> method.
- * A class is considered unloaded if its class loader is finalized.
- * If there no finalization of class loader detected for some timeout,
- * class is considered still loaded and method returns <i>false</i>.
+ * A class is considered unloaded if its class loader is reclaimed.
+ * A runnable lambda is registered to JVM to be called whenever the
+ * class loader becomes unreachable.
+ * After setting the class loader to null, if within a timeout no notification
+ * of its reclaim is received, class is considered still loaded and
+ * <code>unloadClass()</code> method returns <i>false</i>.
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java:52]: All of the existing commentary in this test is somewhat lacking in terms of grammatical correctness, but we can at least make improvements to the new text. I suggest the following:
```
 * A class is eligible for unloading if its class loader has been reclaimed.
 * A Cleaner is used to inform the main test code when the class loader
 * becomes unreachable and is reclaimed.
 * If, after setting the class loader to null, no notification that it has become
 * reclaimed is received within the timeout interval, then the class is considered
 * to still be loaded and <code>unloadClass()</code> returns <i>false</i>.
```

File: test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java [hunk 5]
```diff
@@ -73,19 +77,19 @@ public class ClassUnloader {
     public static final String INTERNAL_CLASS_LOADER_NAME = "nsk.share.CustomClassLoader";
 
     /**
-     * Whole amount of time in milliseconds to wait for class loader finalization.
+     * Whole amount of time in milliseconds to wait for class loader reclaim.
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java:80]: s/reclaim/to be reclaimed/

File: test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java [hunk 6]
```diff
@@ -73,19 +77,19 @@ public class ClassUnloader {
...
      */
     private static final int WAIT_TIMEOUT = 15000;
 
     /**
-     * Piece of time in milliseconds to wait in a loop for class loader finalization.
+     * Piece of time in milliseconds to wait in a loop for class loader reclaim.
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java:85]: Suggestion:
```
* Sleep time in milliseconds for the loop waiting for the class loader to be reclaimed.
* ```

File: test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java [hunk 7]
```diff
@@ -73,19 +77,19 @@ public class ClassUnloader {
...
      */
     private static final int WAIT_DELTA = 1000;
 
     /**
-     * Has class loader been finalized or not.
+     * Has class loader been is_reclaimed or not.
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java:90]: s/is_reclaimed/reclaimed/

File: test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java [hunk 8]
```diff
@@ -131,9 +135,12 @@ public Class<?> getLoadedClass(int index) {
      * @see #setClassLoader(CustomClassLoader)
      */
     public CustomClassLoader createClassLoader() {
-        customClassLoader = new CustomClassLoader(this);
+        customClassLoader = new CustomClassLoader();
         classObjects.removeAllElements();
 
+        // When customClassLoader becomes unreachable, the lambda is called by JVM/GC.
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java:141]: Suggestion:
```
// Register a Cleaner to inform us when the class loader has been reclaimed.
```

File: test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java [hunk 9]
```diff
@@ -146,7 +153,9 @@ public CustomClassLoader createClassLoader() {
     public void setClassLoader(CustomClassLoader customClassLoader) {
         this.customClassLoader = customClassLoader;
         classObjects.removeAllElements();
-        customClassLoader.setClassUnloader(this);
+
+        // When customClassLoader becomes unreachable, the lambda is called by JVM/GC.
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java:157]: Suggestion:
```
// Register a Cleaner to inform us when the class loader has been reclaimed.
```

File: test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java [hunk 10]
```diff
@@ -260,12 +269,12 @@ public boolean unloadClass(ExecutionController stresser) {
...
 
         // force GC to unload marked class loader and its classes
-        if (finalized) {
+        if (is_reclaimed) {
             Runtime.getRuntime().gc();
             return true;
         }
 
-        // class loader has not been finalized
+        // class loader has not been is_reclaimed
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/vmTestbase/nsk/share/ClassUnloader.java:277]: s/is_reclaimed/reclaimed/

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
This looks much simpler/cleaner - thanks!

Just a few nits with comments. Nice try with replacing "finalize" by "reclaim" but it doesn't quite work grammatically in the comments.

**Comment by afshin-zafari** [comment]
> This looks much simpler/cleaner - thanks!
> 
> Just a few nits with comments. Nice try with replacing "finalize" by "reclaim" but it doesn't quite work grammatically in the comments.

Thanks for your comments, and thank you also for being patient with my naive mistakes.

**Review by coleenp — APPROVED** [review]
This looks great.

**Review by dholmes-ora — APPROVED** [review]
Looks good. Thanks for fixing this.

**Comment by afshin-zafari** [comment]
Thanks for reviewing this PR.
/integrate

**Pull request closed without merging**